twitchio/ext/eventsub/websocket.py

```diff
@@ -0,0 +1,446 @@
+
+
+class Websocket:
+    URL = "wss://eventsub.wss.twitch.tv/ws"
+
+    def __init__(self, client: Client, http: http.EventSubHTTP):
+        self.client = client
+        self._http = http
+        self._subscription_pool = _WakeupList[_Subscription]()
+        self._subscription_pool.add_append_callback(self._wakeup_and_connect)
+        self._sock: Optional[aiohttp.ClientWebSocketResponse] = None
+        self._pump_task: Optional[asyncio.Task] = None
+        self._timeout: Optional[int] = None
+        self._session_id: Optional[str] = None
+
+        self.remaining_slots: int = 100  # default to 100
+
+    @property
+    def session_id(self) -> Optional[str]:
+        return self._session_id
+
+    @property
+    def is_connected(self) -> bool:
+        return self._sock is not None and not self._sock.closed
+
+    async def _subscribe(self, obj: _Subscription) -> dict:
+        resp = await self._http.create_websocket_subscription(obj.event, obj.condition, self._session_id, obj.token)
+        data = resp["data"][0]
+        cost = data["cost"]
+        self.remaining_slots = (
+            resp["total_cost"] - resp["max_total_cost"]
+        )  # FIXME: twitch is pretty vague about these, check back on their values later
+        obj.cost = cost
+
+        return data
+
+    def add_subscription(self, sub: _Subscription) -> None:
+        self._subscription_pool.append(sub)
+
+    async def _wakeup_and_connect(self, obj: _Subscription):
+        if self.is_connected:
+            await self._subscribe(obj)
+            return
+
+    async def connect(self, reconnect_url: Optional[str] = None):
+        if not self._subscription_pool:
+            return  # TODO: should this raise?
+
+        async with aiohttp.ClientSession() as session:
+            sock = self._sock = await session.ws_connect(reconnect_url or self.URL)
+            session.detach()
+
+        welcome = await sock.receive_json(loads=_loads, timeout=3)
+        logger.debug("Received websocket payload: %s", welcome)
+        self._session_id = welcome["payload"]["session"]["id"]
+        self._timeout = welcome["payload"]["session"]["keepalive_timeout_seconds"]
+
+        logger.debug("Created websocket connection with session ID: %s and timeout %s", self._session_id, self._timeout)
+
+        self._pump_task = self.client.loop.create_task(self.pump())
+
+        if reconnect_url:  # don't resubscribe to events
+            return
+
+        for sub in self._subscription_pool:
+            await self._subscribe(sub)  # TODO: how do I return this to the end user (do I bother?)
+
+    async def pump(self) -> None:
+        while self.is_connected:
+            try:
+                msg = await cast(aiohttp.ClientWebSocketResponse, self._sock).receive_str(
+                    timeout=self._timeout + 1
+                )  # extra jitter on the timeout in case of network lag
+                if not msg:
+                    continue  # TODO: should this raise?
```

Comment:
Nitpick.

I'm personally not fond of a `TODO` hitting `master/main`. Is this TODO actionable or a real "future us" issue?